From 699d5448f9a69dbca486797f84e4ed2e4519b452 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Thu, 5 Aug 2021 20:06:12 +0200
Subject: bridge: Add additional src and dest mcast lookup option

This adds the new src and dest multicast lookup options to bridge
ports. These are not configured on the bridge itself, but on a bridge
port.
---
 include/linux-private/linux/if_link.h |  2 ++
 include/netlink/route/link/bridge.h   |  4 +++-
 lib/route/link/bridge.c               | 18 +++++++++++++++++-
 3 files changed, 22 insertions(+), 2 deletions(-)

diff --git a/include/linux-private/linux/if_link.h b/include/linux-private/linux/if_link.h
index f1c5496..5f815b0 100644
--- a/include/linux-private/linux/if_link.h
+++ b/include/linux-private/linux/if_link.h
@@ -336,6 +336,8 @@ enum {
 	IFLA_BRPORT_ISOLATED,
 	IFLA_BRPORT_BACKUP_PORT,
 	IFLA_BRPORT_LEARNING_LIMIT,
+	IFLA_BRPORT_DEST_LOOKUP,
+	IFLA_BRPORT_SRC_LOOKUP,
 	__IFLA_BRPORT_MAX
 };
 #define IFLA_BRPORT_MAX (__IFLA_BRPORT_MAX - 1)
diff --git a/include/netlink/route/link/bridge.h b/include/netlink/route/link/bridge.h
index 8b86702..ffbcae3 100644
--- a/include/netlink/route/link/bridge.h
+++ b/include/netlink/route/link/bridge.h
@@ -7,7 +7,7 @@
  *	of the License.
  *
  * Copyright (c) 2013 Thomas Graf <tgraf@suug.ch>
- * Copyright (c) 2020 MaxLinear, Inc.
+ * Copyright (c) 2020-2021 MaxLinear, Inc.
  */
 
 #ifndef NETLINK_LINK_BRIDGE_H_
@@ -42,6 +42,8 @@ enum rtnl_link_bridge_flags {
 	RTNL_BRIDGE_UNICAST_FLOOD	= 0x0010,
 	RTNL_BRIDGE_LEARNING		= 0x0020,
 	RTNL_BRIDGE_LEARNING_SYNC	= 0x0040,
+	RTNL_BRIDGE_DEST_LOOKUP		= 0x0080,
+	RTNL_BRIDGE_SRC_LOOKUP		= 0x0100,
 };
 
 #define RTNL_BRIDGE_HWMODE_VEB BRIDGE_MODE_VEB
diff --git a/lib/route/link/bridge.c b/lib/route/link/bridge.c
index 0aa9df7..c5231a1 100644
--- a/lib/route/link/bridge.c
+++ b/lib/route/link/bridge.c
@@ -7,7 +7,7 @@
  *	of the License.
  *
  * Copyright (c) 2010-2013 Thomas Graf <tgraf@suug.ch>
- * Copyright (c) 2020 MaxLinear, Inc.
+ * Copyright (c) 2020-2021 MaxLinear, Inc.
  */
 
 /**
@@ -122,6 +122,8 @@ static struct nla_policy br_attrs_policy[IFLA_BRPORT_MAX+1] = {
 	[IFLA_BRPORT_LEARNING_SYNC]	= { .type = NLA_U8 },
 	[IFLA_BRPORT_LEARNING_LIMIT]	= { .type = NLA_S32 },
 	[IFLA_BRPORT_UNICAST_FLOOD]	= { .type = NLA_U8 },
+	[IFLA_BRPORT_DEST_LOOKUP]	= { .type = NLA_U8 },
+	[IFLA_BRPORT_SRC_LOOKUP]	= { .type = NLA_U8 },
 };
 
 static void check_flag(struct rtnl_link *link, struct nlattr *attrs[],
@@ -185,6 +187,10 @@ static int bridge_parse_protinfo(struct rtnl_link *link, struct nlattr *attr,
 	check_flag(link, br_attrs, IFLA_BRPORT_LEARNING, RTNL_BRIDGE_LEARNING);
 	check_flag(link, br_attrs, IFLA_BRPORT_LEARNING_SYNC,
 	           RTNL_BRIDGE_LEARNING_SYNC);
+	check_flag(link, br_attrs, IFLA_BRPORT_DEST_LOOKUP,
+	           RTNL_BRIDGE_DEST_LOOKUP);
+	check_flag(link, br_attrs, IFLA_BRPORT_SRC_LOOKUP,
+	           RTNL_BRIDGE_SRC_LOOKUP);
 
 	return 0;
 }
@@ -300,6 +306,14 @@ static int bridge_fill_pi(struct rtnl_link *link, struct nl_msg *msg,
 			NLA_PUT_U8(msg, IFLA_BRPORT_LEARNING_SYNC,
 			           bd->b_flags & RTNL_BRIDGE_LEARNING_SYNC);
 		}
+		if (bd->b_flags_mask & RTNL_BRIDGE_DEST_LOOKUP) {
+			NLA_PUT_U8(msg, IFLA_BRPORT_DEST_LOOKUP,
+			           bd->b_flags & RTNL_BRIDGE_DEST_LOOKUP);
+		}
+		if (bd->b_flags_mask & RTNL_BRIDGE_SRC_LOOKUP) {
+			NLA_PUT_U8(msg, IFLA_BRPORT_SRC_LOOKUP,
+			           bd->b_flags & RTNL_BRIDGE_SRC_LOOKUP);
+		}
 	}
 
 	if (bd->ce_mask & BRIDGE_ATTR_COST)
@@ -862,6 +876,8 @@ static const struct trans_tbl bridge_flags[] = {
 	__ADD(RTNL_BRIDGE_UNICAST_FLOOD,	flood),
 	__ADD(RTNL_BRIDGE_LEARNING,			learning),
 	__ADD(RTNL_BRIDGE_LEARNING_SYNC,	learning_sync),
+	__ADD(RTNL_BRIDGE_DEST_LOOKUP,		dest_lookup),
+	__ADD(RTNL_BRIDGE_SRC_LOOKUP,		src_lookup),
 };
 
 /**
